src/mlpack/methods/reinforcement_learning/ppo_impl.hpp

```diff
@@ -0,0 +1,279 @@
+  PolicyType,
+  ReplayType
+>::Update()
+{
+  // Sample from previous experience.
+  arma::mat sampledStates;
+  std::vector<ActionType> sampledActions;
+  arma::colvec sampledRewards;
+  arma::mat sampledNextStates;
+  arma::icolvec isTerminal;
+
+  replayMethod.Sample(sampledStates, sampledActions, sampledRewards,
+      sampledNextStates, isTerminal);
+
+  arma::rowvec discountedRewards(sampledRewards.n_rows);
+  arma::mat nextActionValues;
+  double values = 0.0;
+  criticNetwork.Predict(sampledNextStates, nextActionValues);
+  for (size_t i = sampledRewards.n_cols; i > 0; --i)
+  {
+    values = sampledRewards[i - 1] + values * config.Discount();
+    discountedRewards[sampledRewards.n_cols - i] = values;
+  }
+
+  arma::mat actionValues, advantages, criticGradients, actorGradients;
+  criticNetwork.Forward(sampledStates, actionValues);
+
+  advantages = arma::conv_to<arma::mat>::
+      from(discountedRewards) - actionValues;
+
+  // Update the critic.
+  criticNetwork.Backward(advantages, criticGradients);
+  criticUpdater.Update(criticNetwork.Parameters(), config.StepSize(),
+      criticGradients);
+
+  for (size_t step = 0; step < config.ActorUpdateStep(); step ++) {
+    // calculate the ratio.
+    arma::mat actionParameter, sigma, mu;
+    actorNetwork.Forward(sampledStates, actionParameter);
+
+    ann::TanhFunction::Fn(actionParameter.row(0), mu);
+    ann::SoftplusFunction::Fn(actionParameter.row(1), sigma);
+
+    ann::NormalDistribution normalDist =
+        ann::NormalDistribution(vectorise(mu, 0), vectorise(sigma, 0));
+
+    oldActorNetwork.Forward(sampledStates, actionParameter);
+    ann::TanhFunction::Fn(actionParameter.row(0), mu);
+    ann::SoftplusFunction::Fn(actionParameter.row(1), sigma);
+
+    ann::NormalDistribution oldNormalDist =
+        ann::NormalDistribution(vectorise(mu, 0), vectorise(sigma, 0));
+
+    // Update the actor.
+    // observation use action.
+    arma::vec prob, oldProb;
+    arma::colvec observation(sampledActions.size());
+    for (size_t i = 0; i < sampledActions.size(); i++) {
+      observation[i] = sampledActions[i].action;
+    }
+    normalDist.LogProbability(observation, prob);
+    oldNormalDist.LogProbability(observation, oldProb);
+
+    arma::mat ratio = arma::exp(vectorise(prob - oldProb, 1));
+
+    arma::mat surrogateLoss = arma::clamp(ratio, 1 - config.Epsilon(),
+        1 + config.Epsilon()) % advantages;
+    arma::mat loss = -arma::min(ratio % advantages, surrogateLoss);
+
+    // backward the gradient
+    arma::mat dratio1 = -loss % (ratio % advantages <= surrogateLoss)
+                        % advantages;
+    arma::mat dsurro = -loss % (ratio % advantages >= surrogateLoss);
+    arma::mat dratio2 = (ratio >= (1 - config.Epsilon())) %
+        (ratio <= (1 + config.Epsilon())) % advantages % dsurro;
+
+    arma::mat dprob = (dratio1 + dratio2) %
+        arma::exp(vectorise(prob - oldProb, 1));
+
+    arma::mat dmu = (vectorise(observation, 1) - mu) /
```

Comment:
See comment above, if you need a rowvec, it's more efficient to just transpose the colvec.